blob/service.go

```diff
@@ -347,127 +354,204 @@ func (s *Service) retrieve(
+			}
+			rowsWithParityShares[rowIndex-inclusiveNamespaceStartRowIndex] = append(
+				rowsWithParityShares[rowIndex-inclusiveNamespaceStartRowIndex],
+				appShare[0],
+			)
 		}
-		getSharesSpan.SetStatus(codes.Error, err.Error())
-		return nil, nil, err
 	}
 
 	getSharesSpan.SetStatus(codes.Ok, "")
 	getSharesSpan.AddEvent("received shares", trace.WithAttributes(
-		attribute.Int64("eds-size", int64(len(header.DAH.RowRoots)))))
+		attribute.Int64("eds-size", int64(squareSize*2))))
 
-	var (
-		appShares = make([]shares.Share, 0)
-		proofs    = make(Proof, 0)
-	)
-
-	for _, row := range namespacedShares {
-		if len(row.Shares) == 0 {
-			// the above condition means that we've faced with an Absence Proof.
-			// This Proof proves that the namespace was not found in the DAH, so
-			// we can return `ErrBlobNotFound`.
-			return nil, nil, ErrBlobNotFound
+	// go over the shares until finding the requested blobs
+	for currentShareIndex := 0; currentShareIndex < len(rowsShares); {
+		currentShareApp, err := shares.NewShare(rowsShares[currentShareIndex])
+		if err != nil {
+			return nil, nil, err
 		}
 
-		appShares, err = toAppShares(row.Shares...)
+		// skip if it's a padding share
+		isPadding, err := currentShareApp.IsPadding()
 		if err != nil {
 			return nil, nil, err
 		}
+		if isPadding {
+			currentShareIndex++
+			continue
+		}
+		isCompactShare, err := currentShareApp.IsCompactShare()
+		if err != nil {
+			return nil, nil, err
+		}
+		if isCompactShare {
+			currentShareIndex++
+			continue
+		}
+		isSequenceStart, err := currentShareApp.IsSequenceStart()
+		if err != nil {
+			return nil, nil, err
+		}
+		if isSequenceStart {
+			// calculate the blob length
+			sequenceLength, err := currentShareApp.SequenceLen()
+			if err != nil {
+				return nil, nil, err
+			}
+			blobLen := shares.SparseSharesNeeded(sequenceLength)
 
-		proofs = append(proofs, row.Proof)
-		index := row.Proof.Start()
-
-		for {
-			var (
-				isComplete bool
-				shrs       []shares.Share
-				wasEmpty   = sharesParser.isEmpty()
-			)
-
-			if wasEmpty {
-				// create a parser if it is empty
-				shrs, err = sharesParser.set(rowIndex*len(header.DAH.RowRoots)+index, appShares)
-				if err != nil {
-					if errors.Is(err, errEmptyShares) {
-						// reset parser as `skipPadding` can update next blob's index
-						sharesParser.reset()
-						appShares = nil
-						break
-					}
-					return nil, nil, err
-				}
-
-				// update index and shares if padding shares were detected.
-				if len(appShares) != len(shrs) {
-					index += len(appShares) - len(shrs)
-					appShares = shrs
-				}
+			endShareIndex := currentShareIndex + blobLen
+			if endShareIndex > len(rowsShares) {
+				// this blob spans to the next row which has a namespace > requested namespace.
+				// this means that we can stop.
+				return nil, nil, ErrBlobNotFound
+			}
+			// convert the blob shares to app shares
+			blobShares := rowsShares[currentShareIndex:endShareIndex]
```

Comment:
should be exclusive, no? since `endShareIndex := currentShareIndex + blobLen`

Reply:
yes, the `endShareIndex` is exclusive. Renamed in: 500f0b8

For example, we have the following row:

0 1 2 3 4 5 6 7

currentShareIndex := 2
blobLen := 3

since `currentShareIndex` is the first share of the blob, the end share of the blob is `4` : `{2, 3, 4}`

2 + 3 = 5 which is exclusive

Reply:
As it exclusive, shouldn't it be:
`blobShares := rowsShares[currentShareIndex:endShareIndex-1]`
currentShareIndex := 2
blobLen := 3
2+3 =5 
But the actual end share is 4

Reply:
for example: `slice := []int{0, 1, 2, 3, 4}`
`slice[0:2] =  {0, 1}`.

The end range is not included in the go slice. Go slicing is end exclusive